color harmony guide lines in RYB vectorscope

Mark-64 opened the pull request:
Fixes #13247 

This PR introduces in RYB vectorscope the visualization of guide lines corresponding to various "color harmonies". See the linked issue for some background.
It is still a draft, for testing and evaluation purposes.

**Use case**
This workflow is inspired by Johanna Kustra's videos on YT about color harmonies.

1) With the global color picker, take live samples of the key color areas of the image. The choice of those is purely artistic, depending on the feeling you want to convey. Display them on the RYB vectorscope.

2) With the new color harmony selection button (the one that replaces the green button, icon is a random one, still to be done), select (by clicking multiple times) the color harmony you want to implement (in the range of doable), or the closest to the distribution of the live samples. Scroll to rotate. Some sectors are longer that others, those are the primary colors.

3) If  some of the samples fall outside the guide lines, bring them inside. You can use any DT color manipulation tool you like. Personally, I like to use color balance RGB + masks (waiting for the future color equalizer module). Use hue shift / croma / saturation to move the targeted live sample in the vectorscope. Remember that vectorscope only shows hue and saturation, not luminance. Keep an eye that the result looks credible, don't try to force fitting the color harmony at any cost, approximate is ok.
Also consider the overall balance of the colors; primary color should be more prominent, meaning more saturated and/or take more space.

As for every tool in photography, the above will not turn a bad picture in a masterpiece, and surely you could judge the color harmony by eyeballs. All in all those are just guidelines, like the ones we have for composition (rule of thirds or similar). Here we try to help the composition of colors. Also, rules are made to be broken sometimes...

Examples (pics are taken from [Joanna Kustra website](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiHkd-LmK78AhXogP0HHaBSDksQFnoECA4QAQ&url=https%3A%2F%2Fjoannakustra.com%2F&usg=AOvVaw0qcLZzGFicAKEj8VK9IkFD))

**1. Monochromatic**

![mono1](https://user-images.githubusercontent.com/43290988/210584971-c008cc72-84fe-401e-b6a5-51b09e7a1b11.png)

![mono12](https://user-images.githubusercontent.com/43290988/210585034-e58f3791-ebd4-46b4-b7f9-d2405417fe3a.png)

**2. Analogous**
![analog2](https://user-images.githubusercontent.com/43290988/210585869-ca9596d3-45fc-4b4d-9602-65a5691aeed8.png)

![analog22](https://user-images.githubusercontent.com/43290988/210585893-199d202c-f2d5-4df7-979f-e7bb5656e4bf.png)

**3. Complementary**
![compl3](https://user-images.githubusercontent.com/43290988/210585588-6815b1f3-03c0-441f-9022-699376f9a78b.png)

![compl32](https://user-images.githubusercontent.com/43290988/210585630-7b1c5243-562a-4159-a7fa-11f864f80a24.png)

**4. Dyad**
![dyad1](https://user-images.githubusercontent.com/43290988/210586090-8061f727-a90e-4900-abc7-b8c5c9602bad.png)

![dyad12](https://user-images.githubusercontent.com/43290988/210586106-b4571abe-7c17-4ee5-9266-0469a2147c17.png)

**5. Split Complementary**
![splitc1](https://user-images.githubusercontent.com/43290988/210586231-59f11183-5908-4937-a59b-0c026699d781.png)

![splitc12](https://user-images.githubusercontent.com/43290988/210586243-ba9f55d3-d343-4010-b6a0-74124744bc46.png)

**6. Triad**
![triad1](https://user-images.githubusercontent.com/43290988/210586425-d2d78e90-a0bd-4ac8-b47e-530aa200dca6.png)

![triad12](https://user-images.githubusercontent.com/43290988/210586473-4836fa51-7373-48b3-9c96-3a8459dafed3.png)

**7. Tetrad**
![tetrad4](https://user-images.githubusercontent.com/43290988/210586586-5cb250f7-5705-46b8-83d6-b88dec71ee65.png)

![tetrad42](https://user-images.githubusercontent.com/43290988/210587599-856c23aa-b656-4f19-ac53-9ba21a8d3c69.png)

Any feedback is welcome

Mark-64 commented:
Fixes #13247 

This PR introduces in RYB vectorscope the visualization of guide lines corresponding to various "color harmonies". See the linked issue for some background.
It is still a draft, for testing and evaluation purposes.

**Use case**
This workflow is inspired by Johanna Kustra's videos on YT about color harmonies.

1) With the global color picker, take live samples of the key color areas of the image. The choice of those is purely artistic, depending on the feeling you want to convey. Display them on the RYB vectorscope.

2) With the new color harmony selection button (the one that replaces the green button, icon is a random one, still to be done), select (by clicking multiple times) the color harmony you want to implement (in the range of doable), or the closest to the distribution of the live samples. Scroll to rotate. Some sectors are longer that others, those are the primary colors.

3) If  some of the samples fall outside the guide lines, bring them inside. You can use any DT color manipulation tool you like. Personally, I like to use color balance RGB + masks (waiting for the future color equalizer module). Use hue shift / croma / saturation to move the targeted live sample in the vectorscope. Remember that vectorscope only shows hue and saturation, not luminance. Keep an eye that the result looks credible, don't try to force fitting the color harmony at any cost, approximate is ok.
Also consider the overall balance of the colors; primary color should be more prominent, meaning more saturated and/or take more space.

As for every tool in photography, the above will not turn a bad picture in a masterpiece, and surely you could judge the color harmony by eyeballs. All in all those are just guidelines, like the ones we have for composition (rule of thirds or similar). Here we try to help the composition of colors. Also, rules are made to be broken sometimes...

Examples (pics are taken from [Joanna Kustra website](https://www.google.com/url?sa=t&rct=j&q=&esrc=s&source=web&cd=&cad=rja&uact=8&ved=2ahUKEwiHkd-LmK78AhXogP0HHaBSDksQFnoECA4QAQ&url=https%3A%2F%2Fjoannakustra.com%2F&usg=AOvVaw0qcLZzGFicAKEj8VK9IkFD) 

**1. Monochromatic**

![mono1](https://user-images.githubusercontent.com/43290988/210584971-c008cc72-84fe-401e-b6a5-51b09e7a1b11.png)

![mono12](https://user-images.githubusercontent.com/43290988/210585034-e58f3791-ebd4-46b4-b7f9-d2405417fe3a.png)

**2. Analogous**
![analog2](https://user-images.githubusercontent.com/43290988/210585869-ca9596d3-45fc-4b4d-9602-65a5691aeed8.png)

![analog22](https://user-images.githubusercontent.com/43290988/210585893-199d202c-f2d5-4df7-979f-e7bb5656e4bf.png)

**3. Complementary**
![compl3](https://user-images.githubusercontent.com/43290988/210585588-6815b1f3-03c0-441f-9022-699376f9a78b.png)

![compl32](https://user-images.githubusercontent.com/43290988/210585630-7b1c5243-562a-4159-a7fa-11f864f80a24.png)

**4. Dyad**
![dyad1](https://user-images.githubusercontent.com/43290988/210586090-8061f727-a90e-4900-abc7-b8c5c9602bad.png)

![dyad12](https://user-images.githubusercontent.com/43290988/210586106-b4571abe-7c17-4ee5-9266-0469a2147c17.png)

**5. Split Complementary**
![splitc1](https://user-images.githubusercontent.com/43290988/210586231-59f11183-5908-4937-a59b-0c026699d781.png)

![splitc12](https://user-images.githubusercontent.com/43290988/210586243-ba9f55d3-d343-4010-b6a0-74124744bc46.png)

**6. Triad**
![triad1](https://user-images.githubusercontent.com/43290988/210586425-d2d78e90-a0bd-4ac8-b47e-530aa200dca6.png)

![triad12](https://user-images.githubusercontent.com/43290988/210586473-4836fa51-7373-48b3-9c96-3a8459dafed3.png)

**7. Tetrad**
![tetrad4](https://user-images.githubusercontent.com/43290988/210586586-5cb250f7-5705-46b8-83d6-b88dec71ee65.png)

![tetrad42](https://user-images.githubusercontent.com/43290988/210587599-856c23aa-b656-4f19-ac53-9ba21a8d3c69.png)

Any feedback is welcome

Mark-64 commented:
For the devs: here below are some questions and some todos in this PR
1) color harmony selection button icon
2) actions: I didn't implement any new one. @dterrahe what should I do? those are kind of black magic to me...
3) @TurboGit Should I put the two new config parameters (harmony type and rotation) in `darktableconfig.xml.in` ?
4) If I try to implement @HSUfineprint [suggestion](https://github.com/darktable-org/darktable/issues/13247#issuecomment-1370411154) , I guess the simplest way is to add another mode in overexposed.c or?

dterrahe commented:
> 2\. actions:

You'd have to add `dt_action_define` and `dt_action_register` lines similar to what's following the other `dtgtk_button_new`s and you'd have to write a patch function to translate a shortcut press into a button click. The action implementation in `histogram.c` isn't the cleanest, also because it appears in both darkroom _and_ tethering.

Without having tried or looked at your PR too closely, since a stack is introduced, you may have to deal with the situation where a button is not visible but the shortcut can still be triggered.

Mark-64 commented:
> You'd have to add

Is that really needed? R/G/B channel buttons, as well as the change vectorscope color space button do not have actions as far as I can see, so why the color harmony button, which is even lower level, should have it?
Or am I wrong?

dterrahe commented:
> Is that really needed?

Oh, sorry; I misinterpreted your question. I thought you _wanted_ to implement actions, so that you or your users could assign a shortcut or trigger it from lua. If _you_ don't care, than _I_ certainly don't care :-) As you say, some of the other buttons don't have associated actions either and I haven't yet received any complaints about that...

TurboGit commented:
> Is that really needed? R/G/B channel buttons, as well as the change vectorscope color space button do not have actions as far as I can see, so why the color harmony button, which is even lower level, should have it?
> Or am I wrong?

I think it would be nice to add actions as this will permit to change using keys and possibly map this to external devices. A general comment, that's not because something is not done that it is correct :) The support for actions is recent, a huge work from @dterrahe and certainly we want to extend that when it makes senses.

TurboGit commented:
> 3\. @TurboGit Should I put the two new config parameters (harmony type and rotation) in `darktableconfig.xml.in` ?

Yes, that's mandatory to ensure that the states are recorded and properly restored across sessions. See for example recent fix I've done, see issue #13263.

Mark-64 commented:
> that's mandatory to ensure that the states are recorded and properly restored across sessions

I will do it, but out of curiosity, so far I have used 
`dt_conf_set_string()` `dt_conf_get_string_const()` 
`dt_conf_get_int()` `dt_conf_set_int()`
for the two parameters and the status gets restored across sessions nevertheless, so what's the purpose of putting them in `darktableconfig.xml.in`?

TurboGit commented:
Ok, indeed those are new so it gets recovered across session but if the user's darktablerc it won't be sanitized. That's a very good point for having entries in `darktableconfig.xml.in`. If the value gets corrupted or incorrectly changed manually the specified default value will be used.

Mark-64 commented:
Actually I am thinking to change the automation and only use scroll, without a new button:
`SCROLL` to rotate
`SHIFT+SCROLL` to fine tune rotate 
`ALT+SCROLL` to switch color harmony
`ALT+SHIFT+SCROLL` to switch between different sector widths: normal (30°), large (45°), line (0°)
Having lines (with no vectorscope dimming) instead of sectors could be useful for other purposes too. For example with "monochromatic" + line we actually have a customizable reference hue line. Many have asked for this in the past, can be useful to quickly match hues between different pictures, not necessarily from the same shoot, or even the same photographer (i.e. taking "inspiration" from others ;-)).

`CTRL+SCROLL` is left to resize the histogram scope

HSUfineprint commented:
IMO, an on/off button of color harmony and it's toolstip should be good for most users to aware there is a new scope view. They don't follow every development here. There are two button positions at the right of change coordinate's button. Another position could be harmony method switcher. Scroll controls can still work.

other thought:

1. I'm wondering if user can drag to change sectors' hue angle directly.
2. `ALT+SHIFT+SCROLL` is a little bit too complected for a simple operation, did you consider using "`SCROLL` when mouse hover the sector".
3. Show hue angle (or NCS style "Y20O"/"B80G") while change/rotate hue.

Mark-64 commented:
So I added the new parameter "width":

**Normal**
![image](https://user-images.githubusercontent.com/43290988/211090585-268822c0-2356-454e-90c2-dbb6d99f30ef.png)

**Large**
![image](https://user-images.githubusercontent.com/43290988/211090673-98dd49a9-ce78-4f73-89d4-e940ae9866ff.png)

**Narrow**
![image](https://user-images.githubusercontent.com/43290988/211090721-9f5ea825-2979-46cc-8e95-ff16be31fa3d.png)

**Line**
![image](https://user-images.githubusercontent.com/43290988/211090966-07b95c0f-9173-4c9a-834e-03f0f5fcd5b2.png)


I added info on the color harmony and rotation, on mouse hover
![image](https://user-images.githubusercontent.com/43290988/211091308-00b45a55-f609-412a-b13c-6f72879a6d79.png)

By the way, combining "monochromatic" + "line", we have the customizable hue line that many have asked for. 
![image](https://user-images.githubusercontent.com/43290988/211091514-9bfcfd66-fd68-4e6e-97f5-93ef52de623f.png)
It can be used for quickly matching a color hue between different shots:

1.  go to the source shot and with the global color picker take a sample of the color you want to emulate, visualize it in the RYB vectorscope
2. rotate the line to cross the sample
3. move to the target shot, remove the first sample,
4. go to the selected area you want to match the hue, take another sample
5. use the DT color manipulation tools (for example color balance rgb and make the new sample move on the line

Mark-64 commented:
What is left to do:
1) a proper icon for the new button "color harmony"
2) the new actions (and add the old vectorscope related ones

I would prefer to do those in separate PRs

dterrahe commented:
Would there be a drawback to moving the buttons to the top-left corner? Then the color buttons could be hidden when not in use (without causing the other buttons to jump to the right). And the red button wouldn't need to be repurposed for a different functionality; just separate stack pages would probably work? Or does it only look strange to me, to sometimes have a useless green/blue button sitting there that can't be clicked (and sometimes a red one too).

Mark-64 commented:
> Or does it only look strange to me, to sometimes have a useless green/blue button sitting there that can't be clicked (and sometimes a red one too).

We are at least two...

HSUfineprint commented:
I'm not an UI designer, it's just my thought of how to show the hue angle.

**traditional angle**
![hue angle](https://user-images.githubusercontent.com/109933528/211128181-53e1d836-198a-412e-ab44-29ae865e04df.png)

**NCS style (0-100 percentage)**
![NCS style](https://user-images.githubusercontent.com/109933528/211128210-f6a2f857-a935-428f-bcf2-3b3ffa452298.png)

**NCS style and 360° scale**
![NCS+angle](https://user-images.githubusercontent.com/109933528/211128240-0f1b9f16-15e1-4233-9984-aab145a64eea.png)

I think that will be help for users in identifying the angle. It contains primary and secondary color hues. NCS (nature color system) is not related to RYB color model. But the mark style of hue value is intuitive and easy to understand.

difrkaguilar commented:
The idea looks good from the design point of view, only a few things should be taken into cue for, the numbering of the angles should be all the time on the outside of the chromatic circle, between it and the black circle, and for that you should include the black circle inside the viewing area of the vectorscope, not leaving out part of it above and below. Because then when you get to positions of 90 and 270 degrees, you would not see the information.

HSUfineprint commented:
> the numbering of the angles should be all the time on the outside of the chromatic circle

I followed the logic of big primary sector / small secondary sectors in split complementary as @Mark-64 designed. I guess other harmony method will not have that big/small style. I believe @Mark-64 will handle the black circle viewing area well if he adopted my proposal.

difrkaguilar commented:
The idea is fine. This is readjusting the RGB icons too and the numbers as I wrote.

![2023-01-07 00 06 15](https://user-images.githubusercontent.com/67672868/211132863-21540b8a-2043-4d58-918d-defde347b0e5.jpg)

HSUfineprint commented:
@difrkaguilar It's concise and looks great. Only one thing to remind, the numbers in NCS hue mark is clockwise. That's why I use "V 43° B" and "B 43° G" instead of "B 43° V" and "G 43° B". So all numbers will have the same rules at any hue sector. The numbers mean distance to the first color. Additionally, maybe we can remove the black circle line for a even cleaner view.

difrkaguilar commented:
> @difrkaguilar It's concise and looks great. Only one thing to remind, the numbers in NCS hue is clockwise. That's why I used "V 43° B" and "B 43° G" instead of "B 43° V" and "G 43° B". So all numbers will have the same rules at any hue sector. The numbers mean distance to the first color. Additionally, maybe we can remove the black circle line for a even cleaner view.
> 
> _edit: attach NCS color circle for reference_ ![NCS-01](https://user-images.githubusercontent.com/109933528/211133599-e9742d8f-06e4-42b4-9ecc-2fd421736e56.jpg)

Love that!

HSUfineprint commented:
> remove the initials of the colors and put the angles in the color that corresponds to the position in which they are located

I don't think it works. The color harmony guide is a range of hue by default. If the color sector is across one of any RYBOGV points, it could be hard to identify what the number means. Especially when user uses large sector width. 

I want to replenish why using NCS-like rules for hue angle. A rotation angle or a absolute angle cannot point out specific color. Let's say 191°, or rotate 101°, you must to find it on scope then realize which color it represented. Most people reconize hue angle as HSB or CIELCh defined, the color angle is different with RYB color circle. Using NCS style, rotate 101° is marked as "Y 19° O". I can realize it's a yellow-orange color immediately, and it contains more yellow than orange (41° : 19°). And NCS is familiar for designers who use color harmony more often. Using original NCS 0-100 percentage method could be more clear, but RYB-OGV 6 sectors means divided 360° to 600. I'm worried about un-continued percentage numbers on display. Using 360° scale and 60° in every RYBOGV sector are way more easy.

Mark-64 commented:
We can make different angle coding coexist, we can cycle through them somehow, if it makes sense to have more than one

Mark-64 commented:
I see there is some more work to do to please everyone...
Back to draft

Mark-64 commented:
For the buttons, the best layout IMO is to put the control buttons on the left, so they can grow freely, and the rgb buttons on the right.
Here all the buttons are shown, in reality some are hidden depending on the view.
@dterrahe  @Nilvus  is that ok for you?

![image](https://user-images.githubusercontent.com/43290988/211146442-647e43a9-d662-49fa-902e-3cc94531b916.png)

dterrahe commented:
> For the buttons, the best layout

An alternative would be to change the current "toggle" button for the mode into a row of four choices on the top left and move all mode-dependent options to the top right. This would be a bigger change, so not required for this PR (but if that would be the plan, you could limit the amount of polishing of a different solution here).

Advantages:
- easier to see what the current mode is (the existing tooltip only says what the _next_ mode will be
- directly selectable, so no chance of clicking once too many and having to cycle all around again
- shortcut or lua code can also directly select a mode, which is not possible now unless you know the _current_ mode

Drawbacks:
- takes more space
- ...?

I think the toggle button made sense when there were only two modes, but should have been replaced when more modes were added.

Mark-64 commented:
> An alternative would be to change the current "toggle" button for the mode into a row of four choices on the top left and move all mode-dependent options to the top right.

I see your point, but then the right side buttons would grow right to left to avoid button jumping. So the last button "show color harmony" would be the leftmost of the right group.
I agree we should think a bit and leave this for another PR

Mark-64 commented:
> The idea is fine, or remove the initials of the colors and put the angles in the color that corresponds to the position in which they are located. RGB icons can be hide or disable if in this case they're not to be used.
> 
> ![2023-01-07 00 06 15](https://user-images.githubusercontent.com/67672868/211133300-d494a761-b111-4b0b-8db5-ffecf82db818.png)

This is not the current layout, the black circle is clipped in reality. The hue notation cannot be written outside the circle of primary colors or it will be clipped.
And I am hesitant to use precious additional vertical gui real estate only for this. It could be drawn inside the circle of primaries, all in all it is just an overlay on mouse hover

dterrahe commented:
> to avoid button jumping

The most important button that shouldn't jump when changing mode is the mode button itself. If the others jump, especially when they are on the other side, it doesn't matter too much imho...

Mark-64 commented:
> NCS (nature color system)

I didn't know this color notation. I went a bit into their [site ](https://ncscolour.com/ncs/) and I am not sure it is useful here.
I am not a color expert by any mean, so take my words with a grain of salt.
They use 6 primaries (R, Y, B, G + black and white) to identify any color. Even if we disregard back and white, 4 primaries remain. 
The vectorscope use 6 primaries (or actually 3 primary + 3 secondary colors). I am not sure the primary colors are the same and if there is an accurate relation with the NCS system.
I would like to hear the opinion of a more expert dev when coming to colors. I don't want to implement something that only resembles the NCS system, but it is inaccurate

Mark-64 commented:
> The most important button that shouldn't jump when changing mode is the mode button itself. If the others jump, especially when they are on the other side, it doesn't matter too much imho...

I agree.
So the left group would be actually radio buttons. We don't have that class in DT, as far as I can see.
So definitely for another PR, I should not put too many different things here.

dterrahe commented:
> So the left group would be actually radio buttons. We don't have that class in DT

Agreed; we use groups of toggle buttons (and "manually" keep them in sync)

> I should not put too many different things here.

Definitely! As said, I only brought it up here so that you don't waste too much time on polishing something that gets ripped out in the next PR anyway...

TurboGit commented:
I'am wondering if we should not put the buttons vertically on both sides? There is often more spaces there .

HSUfineprint commented:
> I would like to hear the opinion of a more expert dev when coming to colors. I don't want to implement something that only resembles the NCS system, but it is inaccurate

Using NCS-like hue coding did not mean we are claiming the vector scope is NCS. You don't need accurate, because it's not NCS at all. It's a way to mark hue angle, and the method is inspired by NCS. The idea is using two color names which the hue to mark is located between them, and the number between two color names means the distance to first color. That's the whole story.

I already posted why I think it's a better way than using a traditional angle number. https://github.com/darktable-org/darktable/pull/13275#issuecomment-1374406517

I have several color management certificates, but I will not say I'm an expert in color. Most important, it's just another thinking way of hue. After all, it's your PR. You are the man who can only decide how it works.

Mark-64 commented:
> After all, it's your PR
I don't want to be dismissive, really. My rationale is that any color harmony, by definition, is a fixed relation between colors on the wheel. All the pattern is rotated at once, so there is only one rotation angle, also for a 4-sectors harmony.
I have been using the code in this PR for a while and to me the visualization of the angle in the gui can serve the purpose of recalling the specific position I have used for a certain picture.
Remember, harmony type and rotation are not properties of the picture, they are not saved in the .xmp, so it could make sense to annotate the setup when switching between pictures.
And to me, it is easier to remember a single number, rather that NCS-like  color code.
I might be wrong, please provide a use case for the NCS notation.

Mark-64 commented:
> I'am wondering if we should not put the buttons vertically on both sides? There is often more spaces there .

And that would be the result. I have no strong opinion on this.

![image](https://user-images.githubusercontent.com/43290988/211168895-7ff9c245-650d-4299-bb21-8b1a311be624.png)

![image](https://user-images.githubusercontent.com/43290988/211168869-e58524fa-6ad7-4d51-bf91-83652f348b4e.png)

TurboGit commented:
Thanks for the mock-up, really no strong opinion either.

Mark-64 commented:
> Thanks for the mock-up

That's the real code.

New icon proposed for the color harmony button, a color swatch.

![image](https://user-images.githubusercontent.com/43290988/211187759-6daf1a74-7cea-42a0-8001-3285b9ba0687.png)

elstoc commented:
I suspect we're getting to the point where the histogram is so overloaded with options that we might need to consider replacing the buttons with a separate hamburger/settings menu or another collapsible utility module to manage it.

Mark-64 commented:
Rebased and squashed in fewer commits. I would say this is ready for review.
There might still be things to do: (actions, radio buttons or other gui changes, different color coding etc) but I should not put too many different things in this PR.

difrkaguilar commented:
> I suspect we're getting to the point where the histogram is so overloaded with options that we might need to consider replacing the buttons with a separate hamburger/settings menu or another collapsible utility module to manage it. The latter option could also allow for presets for users who only use a couple of modes.

That would be a phenomenal solution as it would solve a few problems about the number of options and features to display in the scopes in general, as well as other possible features that may appear on the fly. One of the drawbacks is that you always lose some space vertically. We will have to think about several scenarios, what is the base screen resolution of the darktable? Do users generally keep all their modules deployed? In this new version is an advance that the modules are reorganized with an animation to show all its features and that helps a lot to the visualization. IMHO I would rather sacrifice a small space dedicated to a drop down menu than to have so many options on top of the scope display area and others displayed in "on mouse over" events.

dterrahe commented:
> overloaded with options

Preferences/settings are for options where a user would have a long term preference, set it once and forget about it. Sacrificing a whole line to that seems like a waste (it works fine for modules that have headers anyway).

Toggles for the three colors are different in that you would often switch them on/off repeatedly. Having to go into a menu would be a nuisance. Some users might use several modes regularly too. So how many buttons do we have then that might be worth hiding (at a cost)?

Of course I'm looking forward to the next gui mini-war as much as the next man!

Mark-64 commented:
> Toggles for the three colors are different in that you would often switch them on/off repeatedly.

Fully agree, I switch regularly back and forth between "rbg parade" and vectorscope in every edit.
I believe your suggestion of a row (or column) of radio buttons is the best in terms of usability

Nilvus commented:
I've spent some time to read all your comments here. My first reaction was fear that we overload too much histogram. That's the main reason that @HSUfineprint proposal to add angles is a no go for me. Even if it's sounds visually interesting and seems a good idea. Histogram already use quite some processor (and not all users have and want last processors/graphics cards). In such tiny space, having too much options is not a good idea I think.

@Mark-64: did you make some benchmarks to see resources used by your PR?

Just a thing: I'm not against that PR, I find that proposal interesting even if I'm not sure it's really worth the case. It will be if this take only really few resources.

About icons space, vertical solution have one drawback: if histogram is high on shadows or highlights parts, I fear that icons would be above histogram. Having a menu, why not, that would just add one more click to access options. Personally, as I hide histogram and prefer using my eyes, I don't have any more opinion on that.

For CSS, @Mark-64, I will check after having testing that PR on my side.

Mark-64 commented:
> resources used by your PR

Resource consumption is insignificant.
If guide lines are disabled, no extra resource are take, but also if activated, it's just a simple cairo drawing.
When coming to gui cluttering, I kept it to a minimum. The only visible sign upfront is an extra button in the RYB vectorscope (on mouse hover)

TurboGit commented:
> Just a thing: I'm not against that PR, I find that proposal interesting even if I'm not sure it's really worth the case.

Well that's not my view on this, color harmony is a very interesting and important part of photography and having help via the vectorscope is nice.

Nilvus commented:
@Mark-64: there's a minor refresh issue on your PR.  A precise step-by-step:

1. hide histogram with Shift+ctrl+h
2. close darktable
3. launch darktable then open an image on darkroom
4. show histogram again
5. See that no histogram is displayed, only theme backround (dark, grey...)

Going back on lighttable then opening an image, restore histogram. This issue could only be shown if histogram is hidden when darktable is launched.

See issue:
![image](https://user-images.githubusercontent.com/45535283/211219283-e6df1274-fe43-4cc5-972a-1dde0eb1942e.png)

Nilvus reviewed (changes requested):
@Mark-64: another thing, no need to update CSS. Actual CSS code in master allow what you've done on that PR. Even if on histogram there's a not toggle info, this is for red/blue/green buttons. Your color harmony button inherit then for active, checked and hover effects in parts around lines 1650. Just remove your change so on CSS. You could also remove your color-harmony tag that you create for that.

I tested your PR with master CSS code and compare with yours. Didn't see any change.

dterrahe commented:
> there's a minor refresh issue on your PR.

Not a regression in this PR as it also happens on master; this is similar to #13182 so the same fix works; adding
`gtk_widget_show_all(self->widget);` at the end of `gui_init`. (tbh I don't completely understand why reentering darkroom _after_ enabling the histogram works as well; this suggests there is a show_all call somewhere that might cause problems with other modules).

I'll submit that in a separate PR as it should also go into 4.2.1 as well and this PR likely won't :-)

HSUfineprint commented:
> New icon proposed for the color harmony button, a color swatch.
![211187759-6daf1a74-7cea-42a0-8001-3285b9ba0687](https://user-images.githubusercontent.com/109933528/211231585-aaefbe8c-d7bb-4837-b120-9f24a9fbd53b.png)

Lovely! But color swatch icon usually means "spot color" or "named color". It's related to picking or control specific colors from a knowing database, Pantone for example, in many software. Here are some other icons maybe suit color harmony, you can use them freely.

![color harmony button](https://user-images.githubusercontent.com/109933528/211240380-9a51122f-f0d4-4d5a-b53e-60e3c172f327.png)

Mark-64 commented:
> no need to update CSS

You're right. I messed up with a mistake in widget naming, which I fixed in my last commit. Will revert the changes

Mark-64 commented:
> Here are some other icons

Thanks, I like this one the most.
![image](https://user-images.githubusercontent.com/43290988/211253478-f8d6ee1c-5f7c-4137-8ee0-4678e81c1b0d.png)

Translating into actual Cairo source code could be time consuming, so let's see if we have a consensus here

HSUfineprint commented:
> Remember, harmony type and rotation are not properties of the picture, they are not saved in the .xmp, so it could make sense to annotate the setup when switching between pictures.
And to me, it is easier to remember a single number, rather that NCS-like color code.
I might be wrong, please provide a use case for the NCS notation.

@Mark-64 As I already mention in https://github.com/darktable-org/darktable/pull/13275#issuecomment-1374406517.

1. A rotation angle did not clearly represent a specific color. You must find it on scope first, then can realize it.
2. The cognition of hue angle meaning could be vary for different users, and in different HSB/LCh/RYB color model.
3. The harmony sector direction rotates from 12 o'clock, but in most cases, 0° hue angle is at 3 o'clock.

NCS style angle can fix these issue. And user could easily realize what the sector color is, even no need to check the chromatic circle and primary color position.

![harmony hue_rotation degree](https://user-images.githubusercontent.com/109933528/211280925-023cfb79-5eac-4179-8a43-1510231162e6.png)
![harmony hue_NCS degree](https://user-images.githubusercontent.com/109933528/211280956-f1436f63-299e-4385-be49-3aaa5c088600.png)

As for cross images editing, you can still just remember 4 letters (Y12O or V12B in this case) of any one of the sectors. And meaningful letters is easier to remember than unrelated numbers. But it's still your call.

@Nilvus, I removed all angle marks, you can compare with upper images. I don't think the angle letters will overload the vectorscope. But it needs button and action redesign indeed (in other PR).

![harmony hue_none](https://user-images.githubusercontent.com/109933528/211284000-8803b5a7-0587-40dd-b8b7-a0fecb337b97.png)

Nilvus commented:
Thanks @dterrahe for your quick fix (didn't test but I'm sure it's ok).

@HSUfineprint: seems that way to not overload so much and if this do not need more resources than that PR as it is actually, why not. Remains one thing (maybe some users could deal with that anyway): colors do not have same names, so same initials. How to deal with that?

SoupyGit commented:
Nice work on this. Some thoughts from a non dev, for what it's worth.

Remembering one hue angle is easier than multiple, and less clutter, so I prefer that solution. It's also not clear to me why we would need detailed hue angle info. So in @HSUfineprint latest post, that means the first graphic gets my vote. (If you went with naming colours and they all needed different starting letters, than y yellow, o orange, r red, p pink, m magenta, v violet, b blue, t teal, c cyan, a aqua, g green, l lime - but then question is how consistent these names are with other modules in dt, such as color zones).

Buttons horizontal across top look good to me. Better than introducing menu and more clicks. (I tend to use shortcuts for histogram so menu is not a big deal if all items are still accessible with shortcut, but as it stands there's a couple which are not, so menu not as good for them).

Best icon, from choices given by @HSUfineprint , is bottom row, 2nd from left, because visually it is most similar to how the scope and guides actually look. The one @Mark-64 chose would be my 2nd pick, as its close to how the scope looks when one chooses the line option.

dterrahe commented:
> if all items are still accessible with shortcut, but as it stands there's a couple which are not

This is mostly because some buttons change function depending on mode. When that gets cleaned up (buttons simply get (un)hidden when necessary, like everywhere else) completing shortcuts is trivial.

HSUfineprint commented:
@SoupyGit @Nilvus, in RYB vectorscope, we only have 3 primary color R/Y/B, and 3 secondary color O/G/V. They are fix and will not change. So different color names is not an issue here.

Thanks for everyone, I think it's a done deal. NCS style hue mark is not that easy to understand as I thought, at least for developers. Anyway, it's just a proposal of replacing traditional hue angles. I'm happy I could provide some ideas to @Mark-64, who actually wrote all codes. I'm sorry if I delayed your work.

Maybe we should open an issue to discuss the histogram design, buttons and behavior? There are many ideas in this PR. Only color harmony button should have a final confirmation here. Is that okay for you @TurboGit and @Mark-64?

Mark-64 commented:
> Maybe we should open an issue to discuss the histogram design, buttons and behavior? There are many ideas in this PR.

Yes, that's the best way of working. If this PR gets merged, we can then open one or more FR for the remaining things

TurboGit reviewed (approved):
Thanks!

elstoc commented:
@Mark-64 I've finally got round to writing up all of the various changes that have been made to the scopes module over release 4.4 and 4.6 and would appreciate a review when you have time. See [here](https://github.com/darktable-org/dtdocs/blob/master/content/module-reference/utility-modules/shared/scopes.md). You can contact me on the IRC dev channel or submit an issue/PR to dtdocs if you see anything that needs changing.
